release_notes/model.py

```diff
@@ -0,0 +1,228 @@
+
+    def get_identifier(self) -> str:
+        return str(self.pull_request.number)
+
+
+def create_commit_ref(commit: git.Commit) -> CommitReference:
+    return CommitReference(type=_ref_type_commit, commit=commit)
+
+
+def create_pull_request_ref(pull_request: github3.pulls.ShortPullRequest) -> PullRequestReference:
+    return PullRequestReference(type=_ref_type_pull, pull_request=pull_request)
+
+
+@dataclasses.dataclass
+class SourceBlock:
+    category: str
+    target_group: str
+    note_message: str
+
+    def get_identifier(self) -> str:
+        ''' returns a human-readable identifier which can be used e.g. for duplicate checking.
+        does not include line breaks or spaces
+        '''
+        return f'[{self.category}<{self.target_group}>]{self.note_message}' \
+            .lower().strip().replace(' ', '').replace('\n', '')
+
+    def has_content(self) -> bool:
+        ''' checks if there is any content in the source note block
+        '''
+        if self.note_message.strip().lower() == 'none':
+            return False
+        return all(z and z.strip() for z in (self.category, self.target_group, self.note_message))
+
+    def __hash__(self):
+        return hash(self.get_identifier())
+
+    def __eq__(self, other):
+        if isinstance(other, ReleaseNote):
+            return self.__eq__(other.source_block)
+        if isinstance(other, SourceBlock):
+            return hash(other) == hash(self)
+        return False
+
+
+pattern = re.compile(r'\x60{3}(?P<category>\w+)\s+(?P<target_group>\w+)\n(?P<note>.+?)\n\x60{3}',
+                     flags=re.DOTALL | re.IGNORECASE | re.MULTILINE)
+
+
+def iter_source_blocks(content: str) -> typing.Generator[SourceBlock, None, None]:
+    ''' Searches for code blocks in release note notation and returns all found.
+    Only valid note blocks are returned, which means that the format has been followed.
+    However, it does not check if the category / group exists.
+
+    :param content: the content to look for release notes in
+    :return: a list of valid note blocks
+    '''
+    for res in pattern.finditer(content.replace('\r\n', '\n')):
+        try:
+            block = SourceBlock(category=res.group('category'),
+                                target_group=res.group('target_group'),
+                                note_message=res.group('note'))
+            if block.has_content():
+                yield block
+        except IndexError:
+            # group not found, ignore
+            continue
+
+
+@dataclasses.dataclass
+class ReleaseNote:
+    source_commit: Commit  # the commit where the release notes were initially gathered from
+    source_block: SourceBlock
+
+    raw_body: str  # the raw body of the commit / pull request
+    author: typing.Optional[Author]  # the author of the commit / pull request
+    reference: Reference
+
+    source_component: gci.componentmodel.Component = dataclasses.field(compare=False)
```

Comment:
Removed `dataclasses.field(compare=False)`. I re-used this object from the previous implementation, but since `__eq__` is overwritten anyways this won't make any difference :)